### ISSUE TITLE
Tracker for TensorFlow pip package for Arm

### ISSUE THREAD

**Issue by nitins17** [tensorflow/tensorflow#56088]
cc: @learning-to-play @penpornk @yarri-oss @rishikasinha-tf @mseth10 @nSircombe 

This issue tracks the discussion for installing third party TensorFlow packages for Arm/Graviton.

**Comment by mohantym** [comment]
Hi @nitins17 ! Could you please give more details on above feature request?

**Comment by mseth10** [comment]
Created an issue to PyPI support to increase file size limit https://github.com/pypa/pypi-support/issues/1951

**Comment by mseth10** [comment]
The PyPI file size limit is increased to 500MB. The CD pipeline is able to upload wheels to PyPI now. Please find the v2.9.1 ARM64 wheels here: https://pypi.org/project/aws-tensorflow-cpu

**Comment by learning-to-play** [comment]
@mseth10 To be consistent with the other wheels, could you please use the name `tensorflow-<package-name>` (for example, `tensorflow-aws-cpu`) for TensorFlow packages and `tf-nightly-<package-name>` (for example, `tf-nightly-aws-cpu`) for the TensorFlow Nightly package?
@yarri-oss @rishikasinha-tf @nitins17 FYI

**Comment by mseth10** [comment]
> What flags should I use to reproduce those failures?

Hi @milpuz01 , from the CI logs I see the following command being run
```
bazel test --build_tests_only --config=mkl_aarch64 --copt=-mtune=generic --copt=-march=armv8-a --copt=-O3 --copt=-fopenmp --copt=-flax-vector-conversions --linkopt=-lgomp --test_env=TF_ENABLE_ONEDNN_OPTS=1 --test_env=TF2_BEHAVIOR=1 --test_lang_filters=py --define=no_tensorflow_py_deps=true --verbose_failures=true --test_keep_going --local_test_jobs=64 --test_tag_filters=-nopip,-no_pip,-no_oss,-oss_serial,-no_oss_py37,-gpu,-tpu,-benchmark-test,-v1only,-no_aarch64,-requires-gpu -k -- //bazel_pip/tensorflow/... -//bazel_pip/tensorflow/compiler/tf2tensorrt/... -//bazel_pip/tensorflow/compiler/xrt/... -//bazel_pip/tensorflow/core/tpu/... -//bazel_pip/tensorflow/go/... -//bazel_pip/tensorflow/java/... -//bazel_pip/tensorflow/python/integration_testing/... -//bazel_pip/tensorflow/tools/toolchains/... -//bazel_pip/tensorflow/lite/... -//bazel_pip/tensorflow/compiler/mlir/lite/tests:const-fold.mlir.test -//bazel_pip/tensorflow/compiler/mlir/lite/tests:prepare-tf.mlir.test -//bazel_pip/tensorflow/python:nn_grad_test -//bazel_pip/tensorflow/python:dequantize_op_test -//bazel_pip/tensorflow/python:quantized_ops_test -//bazel_pip/tensorflow/python/data/experimental/kernel_tests/service:cross_trainer_cache_test -//bazel_pip/tensorflow/python/eager:forwardprop_test -//bazel_pip/tensorflow/python/framework:node_file_writer_test -//bazel_pip/tensorflow/python/grappler:memory_optimizer_test -//bazel_pip/tensorflow/python/kernel_tests/linalg:linear_operator_householder_test -//bazel_pip/tensorflow/python/kernel_tests/linalg:linear_operator_inversion_test -//bazel_pip/tensorflow/python/kernel_tests/linalg:linear_operator_block_diag_test -//bazel_pip/tensorflow/python/kernel_tests/linalg:linear_operator_block_lower_triangular_test -//bazel_pip/tensorflow/python/kernel_tests/linalg:linear_operator_kronecker_test -//bazel_pip/tensorflow/python/kernel_tests/math_ops:batch_matmul_op_test -//bazel_pip/tensorflow/python/kernel_tests/nn_ops:conv_ops_test -//bazel_pip/tensorflow/python/kernel_tests/nn_ops:conv2d_backprop_filter_grad_test -//bazel_pip/tensorflow/python/kernel_tests/nn_ops:conv3d_backprop_filter_v2_grad_test -//bazel_pip/tensorflow/python/kernel_tests/nn_ops:atrous_conv2d_test -//bazel_pip/tensorflow/python/ops/parallel_for:math_test
```

**Comment by mseth10** [comment]
> @mseth10 To be consistent with the other wheels, could you please use the name `tensorflow-<package-name>` (for example, `tensorflow-aws-cpu`) for TensorFlow packages and `tf-nightly-<package-name>` (for example, `tf-nightly-aws-cpu`) for the TensorFlow Nightly package? @yarri-oss @rishikasinha-tf @nitins17 FYI

Sure, I can update the package names. We would have to apply for a increase of file size limit again with the new package names, so it might take some time to upload a new wheel.

**Comment by mseth10** [comment]
The nightly pipeline seems to be broken today. Earlier only some tests were failing as mentioned by @milpuz01 , but today the build seems to be failing as well. Does anyone have pointers as to what might have caused it (should be one of yesterday's commits) and how to fix it? Thanks! @mihaimaruseac @learning-to-play @nitins17 
https://github.com/tensorflow/tensorflow/actions/runs/2383006475

**Comment by mihaimaruseac** [comment]
The `opt_einsum` thing has been fixed. It's a Python only dependency which should have been installed via `pip install` in CI but it wasn't when I removed it from Bazel.

**Comment by elfringham** [comment]
Hi @milpuz01 those unit tests started to fail with commit https://github.com/tensorflow/tensorflow/commit/8ea5ed0c392b329a3e0481a3f1f7b0ca86821b84
I can only reproduce them inside the manylinux2014 docker container at the moment. Outside of that those tests pass.

**Comment by milpuz01** [comment]
> Hi @milpuz01 , from the CI logs I see the following command being run

Thanks @mseth10. I can confirm that I can reproduce test failings with that command. The reason is that with that command when running tests it cannot find ```tensorflow.core``` as in here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/data/kernel_tests/iterator_test.py#L21. If I remove this ```--define=no_tensorflow_py_deps=true``` then necessary modules are built and tests pass. Is it possible to have CI without that flag or is there alternative way to make missing packages available? (cc: @mihaimaruseac)

@elfringham When running tests using what is currently in master (commit ```d33233```) and without above flag then I am able to make tests pass.

**Comment by elfringham** [comment]
@milpuz01 Linaro has never used that define in its builds.

The failure I see is

`Executing tests from //tensorflow/python/client:session_list_devices_test
-----------------------------------------------------------------------------
WARNING:tensorflow:Please fix your imports. Module tensorflow.python.training.tracking.base has been moved to tensorflow.python.trackable.base. The old module will be deleted in version 2.11.
WARNING:tensorflow:Please fix your imports. Module tensorflow.python.training.checkpoint_management has been moved to tensorflow.python.checkpoint.checkpoint_management. The old module will be deleted in version 2.9.
WARNING:tensorflow:Please fix your imports. Module tensorflow.python.training.tracking.resource has been moved to tensorflow.python.trackable.resource. The old module will be deleted in version 2.11.
Running tests under Python 3.8.13: /tmp/workspace/venv-cp38-cp38/bin/python3
[ RUN      ] SessionListDevicesTest.testInvalidDeviceNumber
INFO:tensorflow:time(__main__.SessionListDevicesTest.testInvalidDeviceNumber): 0.07s
I0526 11:37:13.623998 281473673940064 test_util.py:2458] time(__main__.SessionListDevicesTest.testInvalidDeviceNumber): 0.07s
[       OK ] SessionListDevicesTest.testInvalidDeviceNumber
[ RUN      ] SessionListDevicesTest.testListDevices
INFO:tensorflow:time(__main__.SessionListDevicesTest.testListDevices): 0.0s
I0526 11:37:13.625836 281473673940064 test_util.py:2458] time(__main__.SessionListDevicesTest.testListDevices): 0.0s
[       OK ] SessionListDevicesTest.testListDevices
[ RUN      ] SessionListDevicesTest.testListDevicesClusterSpecPropagation
2022-05-26 11:37:13.670348: I tensorflow/core/distributed_runtime/rpc/grpc_channel.cc:272] Initialize GrpcChannelCache for job localhost -> {0 -> localhost:40157}
2022-05-26 11:37:13.670824: I tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc:438] Started server with target: grpc://localhost:40157
2022-05-26 11:37:13.701357: I tensorflow/core/distributed_runtime/rpc/grpc_channel.cc:272] Initialize GrpcChannelCache for job localhost -> {0 -> localhost:46037}
2022-05-26 11:37:13.701852: I tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc:438] Started server with target: grpc://localhost:46037
2022-05-26 11:37:13.702681: W tensorflow/core/distributed_runtime/rpc/grpc_session.cc:374] GrpcSession::ListDevices will initialize the session with an empty graph and other defaults because the session has not yet been created.
terminate called after throwing an instance of 'std::out_of_range'
  what():  basic_string::replace
Fatal Python error: Aborted

Thread 0x0000ffffb2594c60 (most recent call first):
  File "/root/.cache/bazel/_bazel_root/db210f68f81d95ddcca9ae96b16ed72c/execroot/org_tensorflow/bazel-out/aarch64-opt/bin/tensorflow/python/client/session_list_devices_test.runfiles/org_tensorflow/tensorflow/python/client/session.py", line 741 in list_devices
  File "/root/.cache/bazel/_bazel_root/db210f68f81d95ddcca9ae96b16ed72c/execroot/org_tensorflow/bazel-out/aarch64-opt/bin/tensorflow/python/client/session_list_devices_test.runfiles/org_tensorflow/tensorflow/python/client/session_list_devices_test.py", line 70 in testListDevicesClusterSpecPropagation
  File "/opt/python/cp38-cp38/lib/python3.8/unittest/case.py", line 633 in _callTestMethod
  File "/opt/python/cp38-cp38/lib/python3.8/unittest/case.py", line 676 in run
  File "/opt/python/cp38-cp38/lib/python3.8/unittest/case.py", line 736 in __call__
  File "/opt/python/cp38-cp38/lib/python3.8/unittest/suite.py", line 122 in run
  File "/opt/python/cp38-cp38/lib/python3.8/unittest/suite.py", line 84 in __call__
  File "/opt/python/cp38-cp38/lib/python3.8/unittest/suite.py", line 122 in run
  File "/opt/python/cp38-cp38/lib/python3.8/unittest/suite.py", line 84 in __call__
  File "/opt/python/cp38-cp38/lib/python3.8/unittest/runner.py", line 176 in run
  File "/opt/python/cp38-cp38/lib/python3.8/unittest/main.py", line 271 in runTests
  File "/opt/python/cp38-cp38/lib/python3.8/unittest/main.py", line 101 in __init__
  File "/root/.cache/bazel/_bazel_root/db210f68f81d95ddcca9ae96b16ed72c/execroot/org_tensorflow/bazel-out/aarch64-opt/bin/tensorflow/python/client/session_list_devices_test.runfiles/absl_py/absl/testing/absltest.py", line 2537 in _run_and_get_tests_result
  File "/root/.cache/bazel/_bazel_root/db210f68f81d95ddcca9ae96b16ed72c/execroot/org_tensorflow/bazel-out/aarch64-opt/bin/tensorflow/python/client/session_list_devices_test.runfiles/absl_py/absl/testing/absltest.py", line 2568 in run_tests
  File "/root/.cache/bazel/_bazel_root/db210f68f81d95ddcca9ae96b16ed72c/execroot/org_tensorflow/bazel-out/aarch64-opt/bin/tensorflow/python/client/session_list_devices_test.runfiles/absl_py/absl/testing/absltest.py", line 2156 in _run_in_app
  File "/root/.cache/bazel/_bazel_root/db210f68f81d95ddcca9ae96b16ed72c/execroot/org_tensorflow/bazel-out/aarch64-opt/bin/tensorflow/python/client/session_list_devices_test.runfiles/absl_py/absl/testing/absltest.py", line 2049 in main
  File "/root/.cache/bazel/_bazel_root/db210f68f81d95ddcca9ae96b16ed72c/execroot/org_tensorflow/bazel-out/aarch64-opt/bin/tensorflow/python/client/session_list_devices_test.runfiles/org_tensorflow/tensorflow/python/platform/googletest.py", line 51 in g_main
  File "/root/.cache/bazel/_bazel_root/db210f68f81d95ddcca9ae96b16ed72c/execroot/org_tensorflow/bazel-out/aarch64-opt/bin/tensorflow/python/client/session_list_devices_test.runfiles/absl_py/absl/app.py", line 258 in _run_main
  File "/root/.cache/bazel/_bazel_root/db210f68f81d95ddcca9ae96b16ed72c/execroot/org_tensorflow/bazel-out/aarch64-opt/bin/tensorflow/python/client/session_list_devices_test.runfiles/absl_py/absl/app.py", line 312 in run
  File "/root/.cache/bazel/_bazel_root/db210f68f81d95ddcca9ae96b16ed72c/execroot/org_tensorflow/bazel-out/aarch64-opt/bin/tensorflow/python/client/session_list_devices_test.runfiles/org_tensorflow/tensorflow/python/platform/googletest.py", line 60 in main_wrapper
  File "/root/.cache/bazel/_bazel_root/db210f68f81d95ddcca9ae96b16ed72c/execroot/org_tensorflow/bazel-out/aarch64-opt/bin/tensorflow/python/client/session_list_devices_test.runfiles/org_tensorflow/tensorflow/python/platform/benchmark.py", line 503 in benchmarks_main
  File "/root/.cache/bazel/_bazel_root/db210f68f81d95ddcca9ae96b16ed72c/execroot/org_tensorflow/bazel-out/aarch64-opt/bin/tensorflow/python/client/session_list_devices_test.runfiles/org_tensorflow/tensorflow/python/platform/googletest.py", line 62 in main
  File "/root/.cache/bazel/_bazel_root/db210f68f81d95ddcca9ae96b16ed72c/execroot/org_tensorflow/bazel-out/aarch64-opt/bin/tensorflow/python/client/session_list_devices_test.runfiles/org_tensorflow/tensorflow/python/client/session_list_devices_test.py", line 81 in <module>
`

**Comment by elfringham** [comment]
@mseth10 I see you have used the tag "-no_oss_py37" - can you tell me where that has come from and why you are using it?

**Comment by nSircombe** [comment]
I think this is a bit of a mangling of `-no_oss_py2`, it's set as `-no_oss_py${py_ver}` - I don't think it does anything and is just ignored, but I may be wrong.

**Comment by mihaimaruseac** [comment]
Segfaults arising from C++17 switch are mostly due to `std::string_view(nullptr)`. I'm currently hunting for these and fixing, but will take a while to get to all of them

**Comment by mseth10** [comment]
Requested PyPI support to increase file size limit for these projects:
`tf-nightly-cpu-aws` https://github.com/pypa/pypi-support/issues/1962
`tensorflow-cpu-aws` https://github.com/pypa/pypi-support/issues/1963

**Comment by nitins17** [comment]
> Should we remove `--define=no_tensorflow_py_deps=true` from bazel test command? @milpuz01 @elfringham @mihaimaruseac

We need this flag for the pip tests. 

https://github.com/tensorflow/build/blob/master/tf_sig_build_dockerfiles/devel.usertools/cpu.bazelrc#L58-L66

**Comment by mseth10** [comment]
We got file size limit updated for the updated project name and we have v2.9.1 wheels uploaded there: https://pypi.org/project/tensorflow-cpu-aws/

We are still waiting for an increase in file size limit for the nightly project.

**Comment by nitins17** [comment]
Thanks @mseth10. We are setting up the infrastructure to upload the installer wheels. I will update once we are ready for testing.

**Comment by elfringham** [comment]
https://github.com/tensorflow/tensorflow/pull/56364

**Comment by mseth10** [comment]
https://github.com/tensorflow/tensorflow/pull/56361 required to enable nightly CD pipeline. @nitins17

**Comment by nitins17** [comment]
> @nitins17 Installed the tensorflow wrapper wheel (py38) you shared, but that only installed `tensorflow` package, and not the `tensorflow-cpu-aws` package. Tests also fail:
> 
> ```
> >>> import tensorflow as tf
> >>> t1=tf.constant([1,2,3,4])
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
> AttributeError: module 'tensorflow' has no attribute 'constant'
> ```

Could you run the following and let me know what you get?

```
import platform
print(platform.system())
print(platform.machine())
```

**Comment by mseth10** [comment]
```
>>> import platform
>>> print(platform.system())
Linux
>>> print(platform.machine())
aarch64
```

**Comment by mseth10** [comment]
Opened a GItHub issue to discuss ways to ensure the nightly pipeline passes https://github.com/tensorflow/tensorflow/issues/56377

**Comment by nitins17** [comment]
> @nitins17
> 
> ```
> >>> print(platform.machine())
> aarch64
> ```

Could you try again with these wheels?
- [Python 3.7](https://storage.googleapis.com/tensorflow-nightly/prod/tensorflow/tf_partner_build_installer/release/tensorflow-2.9.1-cp37-cp37m-manylinux_2_17_aarch64.manylinux2014_aarch64.whl)
- [Python 3.8](https://storage.googleapis.com/tensorflow-nightly/prod/tensorflow/tf_partner_build_installer/release/tensorflow-2.9.1-cp38-cp38-manylinux_2_17_aarch64.manylinux2014_aarch64.whl)
- [Python 3.9](https://storage.googleapis.com/tensorflow-nightly/prod/tensorflow/tf_partner_build_installer/release/tensorflow-2.9.1-cp39-cp39-manylinux_2_17_aarch64.manylinux2014_aarch64.whl)
- [Python 3.10](https://storage.googleapis.com/tensorflow-nightly/prod/tensorflow/tf_partner_build_installer/release/tensorflow-2.9.1-cp310-cp310-manylinux_2_17_aarch64.manylinux2014_aarch64.whl)

The issue was that the installer wheel would install the tensorflow-cpu-aws package only if `platform.machine()` returned "arm64". I have [changed the setup.py ](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/pip_package/setup_partner_builds.py#L69-L70)to include "aarch64" as well.

**Comment by mseth10** [comment]
ARM-CD pipeline is uploading nightly wheels to PyPI now https://pypi.org/project/tf-nightly-cpu-aws/#history
It's using `2.10.0` as the version. I'd like to update the version so that it is in this format `2.10.0.dev<yyyymmdd>` same as this project https://pypi.org/project/tf-nightly-cpu/#history
@nitins17 can you point me to the logic in tensorflow repo where this version name format is taken care of.. I will look into how to incorporate that into arm wheels.

**Comment by mseth10** [comment]
@nitins17 tested the Python 3.8 wheel and it works. Installs the `tensorflow-cpu-aws` package. Also ran tests which seem to pass.
```
>>> import tensorflow as tf
>>> t1=tf.constant([1,2,3,4]) 
>>> t2=tf.constant([5,6,7,8])
>>> print(tf.add(t1,t2).shape)
(4,)
>>> 
>>> print(1) if "_v2.keras" in tf.keras.__name__ else print(0)
1
>>> 
>>> print(1) if "_v2.estimator" in tf.estimator.__name__ else print(0)
1
>>> 
```

**Comment by mseth10** [comment]
Thanks! Any particular reason we use `nightly` branch for nightly releases? @nitins17

**Comment by nitins17** [comment]
I think using the nightly branch could be beneficial because:
1. It would be consistent with our x86_64 Linux jobs which use this branch to build its nightlies.
2. If both the x86_64 wheels and aarch64 wheels are built at the same commit, it might be easier to debug in case of failures.

**Comment by mseth10** [comment]
@nitins17 Also, is `nightly` branch synced with `master` branch once a day? What time would that be?

**Comment by nitins17** [comment]
Yes, the `nightly` branch is synced to `master` every evening at around 8pm.

**Comment by mseth10** [comment]
@nitins17 created a PR to use nightly branch and update version https://github.com/tensorflow/tensorflow/pull/56440

**Comment by nSircombe** [comment]
I have a question about the last passing run - why are there only 920 tests? 
With `tensorflow/tools/ci_build/build_scripts/DEFAULT_TEST_TARGETS.sh`, shouldn't we have closer to 3k? Even with a few exclusions and test filters?

**Comment by nitins17** [comment]
> 

Our Linux builds currently run 1358 pip tests. These are the tags/flags that we use: https://github.com/tensorflow/build/blob/master/tf_sig_build_dockerfiles/devel.usertools/cpu.bazelrc#L71-L80

**Comment by nSircombe** [comment]
> Our Linux builds currently run 1358 pip tests. 

Ok, thanks @nitins17 - are these the flags & filters used for all builds, nightly / pre-commit / release etc.?
That's still more than this AArch64 pipeline has run though - is that simply the `-no_aarch64` flag? Or possibly the `-no_oss_py#` flag? (should this one be removed from the AArch64 pipeline - I believe that was suggested at one of the SIG build meetings?).

I have a downstream pipeline running the unit tests on an AArch64 dev build - I have not set `-nopip` or `-no_pip` - should I? It looks like this reduces the total number of cases.

**Comment by nitins17** [comment]
> Ok, thanks @nitins17 - are these the flags & filters used for all builds, nightly / pre-commit / release etc.?

[--config=pip](https://github.com/tensorflow/build/blob/master/tf_sig_build_dockerfiles/devel.usertools/cpu.bazelrc#L71-L80) is for nightly and release. For presubmits, we use [--config=pycpp](https://github.com/tensorflow/build/blob/master/tf_sig_build_dockerfiles/devel.usertools/cpu.bazelrc#L114-L117) and [--config=rbe ](https://github.com/tensorflow/build/blob/master/tf_sig_build_dockerfiles/devel.usertools/cpu.bazelrc#L87-L111)for building faster.

Also, note that we have two suites of tests for nightly and release:
1. "pip" tests which builds the wheel and then runs the tests on the built wheel. 
2. "[nonpip](https://github.com/tensorflow/build/blob/master/tf_sig_build_dockerfiles/devel.usertools/cpu.bazelrc#L49-L56)" tests which test using the source code directly

> That's still more than this AArch64 pipeline has run though - is that simply the -no_aarch64 flag? Or possibly the -no_oss_py# flag? (should this one be removed from the AArch64 pipeline - I believe that was suggested at one of the SIG build meetings?).

I don't think it is because of `-no_aarch64` because a quick search brought up only 3 tests that use this flag.
We don't use the `-no_oss_py*` flag and I can see many tests with this flag so it is possible that the missing tests are due to including this. I would remove this flag and see what we get. 

> I have a downstream pipeline running the unit tests on an AArch64 dev build - I have not set -nopip or -no_pip - should I? It looks like this reduces the total number of cases.

If you are building the wheel and then running the tests on the built wheel, then I would include `-nopip` and `-no_pip` as these are known cases of tests that are currently incompatible with wheel testing.

**Comment by mseth10** [comment]
> Hi @mseth10, Thank you for identifying the commit that caused the breakage. You can send a PR to roll back < 24 hours (ideally within a few hours) after its submission. Because this commit was submitted more than a few days ago, please feel free to send a PR to fix the issue. We can review and submit if it passes all the tests.

Thanks @learning-to-play ! If we can rollback the commit within a few hours of CI failure, then that would be a great step ahead. To make sure we do not miss that, I think it's important if we can automate this step. I came to know that we have this automation in place for x86 already, if so, we can extend it to aarch64 as well(?) Can someone share more details on how this automation works currently? @nitins17

**Comment by elfringham** [comment]
I submitted the PR https://github.com/tensorflow/tensorflow/pull/56620 to remove uses of no_oss_py2, which has been approved but not merged yet, though not sure why.
So it seems that the ARM_CI and ARM_CD workflows are only running the pip tests. Are there plans to also run the nonpip tests on the source code directly?
The current test flags for these ARM workflows include "-requires-gpu,-gpu,-tpu," which are not used in the x86 pip tests.

**Comment by mihaimaruseac** [comment]
Google has been OOO due to 4th of July until today. It will probably get reviewed soon internally

**Comment by mseth10** [comment]
> @mseth10 Could you run `pip install tf-nightly` and verify that it pulls in the `tf-nightly-cpu-aws` package and that all the verification tests pass?

Hi @nitins17 , I verified `pip install tf-nightly` works for python3.8. It installs `tf-nightly-cpu-aws` package and all the verification tests pass.

**Comment by mihaimaruseac** [comment]
So, there is automatic rollback for Google internal tests from projects that are clients of TF. If a test there breaks, the changelist gets rolled back and that results in a rollback commit on GitHub. But, this is only on internal tests, only on x64_64 since that's what we build.

For other cases, I guess we could make a GitHub Action that would run CI from time to time or watch CI results and create a rollback PR. But it has the same drawbacks as any other PR: it needs to be approved, integrated, approved again, run all CI, etc. Sadly, given that TF's source of truth is internal, we cannot do much here (and LLVM story makes it impossible to switch source of truth to external)

**Comment by nitins17** [comment]
@mseth10, @elfringham, @nSircombe Looks like there were no uploads to the `tf-nightly-cpu-aws` project on PyPI for the last few days as the Arm CD workflow has been [failing](https://github.com/tensorflow/tensorflow/actions/workflows/arm-cd.yml) due to a test failure (`tensorflow/python/kernel_tests/nn_ops:pooling_ops_3d_test_cpu`). fyi: @yarri-oss, @rishikasinha-tf

**Comment by mseth10** [comment]
> @mseth10, @elfringham, @nSircombe Looks like there were no uploads to the `tf-nightly-cpu-aws` project on PyPI for the last few days as the Arm CD workflow has been [failing](https://github.com/tensorflow/tensorflow/actions/workflows/arm-cd.yml) due to a test failure (`tensorflow/python/kernel_tests/nn_ops:pooling_ops_3d_test_cpu`). fyi: @yarri-oss, @rishikasinha-tf

I checked the CI [history](https://github.com/tensorflow/tensorflow/actions/workflows/arm-ci.yml?query=event%3Apush), and looks like this commit caused it https://github.com/tensorflow/tensorflow/commit/9178ac9d6389bdc54638ab913ea0e419234d14eb . Added the test to the skip-test list: https://github.com/tensorflow/tensorflow/pull/56829

**Comment by nitins17** [comment]
It looks like #56828 should fix this but we need to wait for it to be reviewed and merged so let's skip the test for now. We can remove it from the skip list after #56828 is merged.

**Comment by ddelange** [comment]
Is this issue a (more active) duplicate of #52973?

**Comment by mihaimaruseac** [comment]
I think it's a little bit more, but that issue should be deduplicated here

**Comment by mseth10** [comment]
> It looks like #56828 should fix this but we need to wait for it to be reviewed and merged so let's skip the test for now. We can remove it from the skip list after #56828 is merged.

@nitins17 removing the test from skip-list and adding a new one as it's been failiing because of a recent update. Please help merge this PR https://github.com/tensorflow/tensorflow/pull/56901

**Comment by nitins17** [comment]
Just a FYI, the Arm CD workflow [failed](https://github.com/tensorflow/tensorflow/runs/7535892666?check_suite_focus=true) yesterday because of an LLVM integrate. I believe https://github.com/tensorflow/tensorflow/commit/47c640a961874f644cd071752835c7b792450bb8 should fix it.

**Comment by mseth10** [comment]
> Just a FYI, the Arm CD workflow [failed](https://github.com/tensorflow/tensorflow/runs/7535892666?check_suite_focus=true) yesterday because of an LLVM integrate. I believe [47c640a](https://github.com/tensorflow/tensorflow/commit/47c640a961874f644cd071752835c7b792450bb8) should fix it.

@nitins17 from CI history it looks like https://github.com/tensorflow/tensorflow/commit/47c640a961874f644cd071752835c7b792450bb8 fixed the issue. But it looks like ARM CD ran on this commit yesterday, and still failed.

**Comment by nitins17** [comment]
From the logs, it looks like Arm CD ran on https://github.com/tensorflow/tensorflow/commit/bc0e3bf0f30a59d25aa1558cec37095db6537762 yesterday which is behind https://github.com/tensorflow/tensorflow/commit/47c640a961874f644cd071752835c7b792450bb8 which fixes the issue. Is it possible to trigger the Arm CD workflow manually at a specific commit?

**Comment by elfringham** [comment]
I have added a new [PR](https://github.com/tensorflow/tensorflow/pull/56933) which removes some tests from the exclude list used by the ARM_CI/ARM_CD actions. These tests currently pass so should be run to keep the test coverage as high as possible.

**Comment by nitins17** [comment]
The [Tensorflow 2.10 branch ](https://github.com/tensorflow/tensorflow/tree/r2.10) cut has been completed at commit `b835ad5009f42980328d973161b362860bcfa451`.

**Comment by mseth10** [comment]
@elfringham I see that we have removed `pad_op_test_{cpu/gpu}` from the arm skip-test list. While these tests pass on master branch, they are [failing](https://github.com/tensorflow/tensorflow/actions/workflows/arm-ci.yml?query=branch%3Ar2.10) on r2.10 branch. Is there a fix on master branch that needs to be backported to r2.10 branch for these tests?

**Comment by mseth10** [comment]
@elfringham I have also tagged you in this PR for review of scripts to run non-pip tests in CI: https://github.com/tensorflow/tensorflow/pull/57113 . Thanks!

**Comment by elfringham** [comment]
> @elfringham I see that we have removed `pad_op_test_{cpu/gpu}` from the arm skip-test list. While these tests pass on master branch, they are [failing](https://github.com/tensorflow/tensorflow/actions/workflows/arm-ci.yml?query=branch%3Ar2.10) on r2.10 branch. Is there a fix on master branch that needs to be backported to r2.10 branch for these tests?

@mseth10 This should have been fixed by https://github.com/tensorflow/tensorflow/commit/c194d8cb4f1b76e1893523ec9bda5829d35bc829#diff-39eadfc771b57ef11bff67b85cc38a47b9c13ec203b145e3feaf556a43bdccb1 AFAIK - but maybe that is not suffficient.
This is the last of the tests that begins to fail with numpy>=1.23.0 but that is only being used by the ARM_Ci/ARM_CD actions. This is because there is no constraint on the version of scipy I think. The x86 CI builds use tensorflow/tools/ci_build/release/requirements_common.txt which limits scipy to 1.7.2 and that in turn limits numpy to < 1.23.0 and this issue is not seen.

**Comment by mseth10** [comment]
> > @elfringham I see that we have removed `pad_op_test_{cpu/gpu}` from the arm skip-test list. While these tests pass on master branch, they are [failing](https://github.com/tensorflow/tensorflow/actions/workflows/arm-ci.yml?query=branch%3Ar2.10) on r2.10 branch. Is there a fix on master branch that needs to be backported to r2.10 branch for these tests?
> 
> @mseth10 This should have been fixed by [c194d8c#diff-39eadfc771b57ef11bff67b85cc38a47b9c13ec203b145e3feaf556a43bdccb1](https://github.com/tensorflow/tensorflow/commit/c194d8cb4f1b76e1893523ec9bda5829d35bc829#diff-39eadfc771b57ef11bff67b85cc38a47b9c13ec203b145e3feaf556a43bdccb1) AFAIK - but maybe that is not suffficient. This is the last of the tests that begins to fail with numpy>=1.23.0 but that is only being used by the ARM_Ci/ARM_CD actions. This is because there is no constraint on the version of scipy I think. The x86 CI builds use tensorflow/tools/ci_build/release/requirements_common.txt which limits scipy to 1.7.2 and that in turn limits numpy to < 1.23.0 and this issue is not seen.

@elfringham I compared ARM CI/CD for master and r2.10 branches. Both install scipy 1.9.0 and numpy 1.23.2 . And both contain the fix that you mentioned. Still only r2.10 fails. Maybe something else is missing. For now we can skip the two tests in r2.10 .

**Comment by elfringham** [comment]
> > > @elfringham I see that we have removed `pad_op_test_{cpu/gpu}` from the arm skip-test list. While these tests pass on master branch, they are [failing](https://github.com/tensorflow/tensorflow/actions/workflows/arm-ci.yml?query=branch%3Ar2.10) on r2.10 branch. Is there a fix on master branch that needs to be backported to r2.10 branch for these tests?
> > 
> > 
> > @mseth10 This should have been fixed by [c194d8c#diff-39eadfc771b57ef11bff67b85cc38a47b9c13ec203b145e3feaf556a43bdccb1](https://github.com/tensorflow/tensorflow/commit/c194d8cb4f1b76e1893523ec9bda5829d35bc829#diff-39eadfc771b57ef11bff67b85cc38a47b9c13ec203b145e3feaf556a43bdccb1) AFAIK - but maybe that is not suffficient. This is the last of the tests that begins to fail with numpy>=1.23.0 but that is only being used by the ARM_Ci/ARM_CD actions. This is because there is no constraint on the version of scipy I think. The x86 CI builds use tensorflow/tools/ci_build/release/requirements_common.txt which limits scipy to 1.7.2 and that in turn limits numpy to < 1.23.0 and this issue is not seen.
> 
> @elfringham I compared ARM CI/CD for master and r2.10 branches. Both install scipy 1.9.0 and numpy 1.23.2 . And both contain the fix that you mentioned. Still only r2.10 fails. Maybe something else is missing. For now we can skip the two tests in r2.10 .

Ah, OK, it looks like this commit is also needed then https://github.com/tensorflow/tensorflow/commit/c1f68309986a4a7105c55545eac92b8b0f5917ab

**Comment by mseth10** [comment]
> Ah, OK, it looks like this commit is also needed then [c1f6830](https://github.com/tensorflow/tensorflow/commit/c1f68309986a4a7105c55545eac92b8b0f5917ab)

@nitins17 @penpornk can you please cherry-pick this commit into r2.10

**Comment by nitins17** [comment]
Cherrypicked https://github.com/tensorflow/tensorflow/commit/c1f68309986a4a7105c55545eac92b8b0f5917ab into [r2.10](https://github.com/tensorflow/tensorflow/commits/r2.10)

**Comment by mseth10** [comment]
@nitins17 https://github.com/tensorflow/tensorflow/pull/57400 created this PR to update r2.9 triggers in accordance with master branch

**Comment by dakl** [comment]
So am I reading this right if I think this will make it possible to install tensorflow on Apple Silicon in docker?

**Comment by nitins17** [comment]
> So am I reading this right if I think this will make it possible to install tensorflow on Apple Silicon in docker?

There is a separate effort currently in progress for making Apple Silicon wheels available for TensorFlow (https://github.com/tensorflow/tensorflow/issues/57185). This issue is for Arm based Linux wheels.

**Comment by dakl** [comment]
> > So am I reading this right if I think this will make it possible to install tensorflow on Apple Silicon in docker?
> 
> 
> 
> There is a separate effort currently in progress for making Apple Silicon wheels available for TensorFlow (https://github.com/tensorflow/tensorflow/issues/57185). This issue is for Arm based Linux wheels. 

That's exactly what I'm after. This will enable tf in Linux-based containers on macOS, since docker for Mac starts a Linux VM under the hood.

**Comment by mseth10** [comment]
grappler remapper tests failing after this commit https://github.com/tensorflow/tensorflow/commit/9d4e950687b25f78fe30a74a8b40ba16f12c9fd1 , is this a known issue and is the fix WIP? @nitins17

**Comment by elfringham** [comment]
There was a new unit test a few days ago on master, //tensorflow/compiler/xla/runtime:custom_call_test that fails on Linaro CI. My testing also shows it failing on x86 so I am not sure why this is not picked up by the main CI testing. I could not find the run that made it pass, it went straight from not existing in the log to having a cached pass.
I have a PR to fix it that also needs an AARCH64 specific part https://github.com/tensorflow/tensorflow/pull/57586

**Comment by elfringham** [comment]
BTW the above fail is actually a fail to build.

**Comment by nitins17** [comment]
> CD on master branch failing on grappler remapper tests after this commit [9d4e950](https://github.com/tensorflow/tensorflow/commit/9d4e950687b25f78fe30a74a8b40ba16f12c9fd1) , is this a known issue and is the fix WIP? @nitins17

This does not seem to be failing on x86.

**Comment by elfringham** [comment]
@nitins17 @angerson PR for manylinux2014_aarch64 docker container with Dual ABI enabled. https://github.com/tensorflow/tensorflow/pull/57634

**Comment by elfringham** [comment]
@cantonios your recent commit introducing FP8 is not building on AARCH64. https://github.com/tensorflow/tensorflow/issues/57918

**Comment by elfringham** [comment]
A new [PR](https://github.com/tensorflow/tensorflow/pull/57969) for an AARCH64 specific issue that was introduced with an update to Eigen with this commit https://github.com/tensorflow/tensorflow/commit/09d49d9f6c9c13679ab8181d1b15480ea8a88da1

**Comment by elfringham** [comment]
My PR https://github.com/tensorflow/tensorflow/pull/57969 was approved 3 days ago but has not yet been merged. Can anyone please see if there is anything that needs to be addressed?
